big_vision/trainers/proj/gsam/gsam.py

```diff
@@ -0,0 +1,94 @@
+import jax
+import jax.numpy as jnp
+
+def dual_vector(y):
+  """Returns the solution of max_x y^T x s.t. ||x||_2 <= 1.
+  Args:
+    y: A pytree of numpy ndarray, vector y in the equation above.
+  """
+  gradient_norm = jnp.sqrt(sum(
+        [jnp.sum(jnp.square(e)) for e in jax.tree_util.tree_leaves(y)]))
+  normalized_gradient = jax.tree_map(lambda x: x / gradient_norm, y)
+  return normalized_gradient, gradient_norm
+
+def gsam_gradient(loss_fn, base_opt, inputs, targets,
+                  rho_max, rho_min, alpha, lr, lr_max, lr_min, eps=1e-12,
+                  adaptive_perturbation=False, minimize_fp=True):
+  """
+  Get the GSAM gradient (https://openreview.net/pdf?id=edONMAnhLu-) of the loss function.
+  Args: 
+    loss_fn: the loss function.
+    base_opt: the base optimizer.
```

Comment:
[1/2] This (`base_opt.target` used below) does not work anymore with optax. Although it looks like you really use `base_opt` only for getting to the params, so you can replace the argument by an actual `params` argument, and then use that everywhere where you currently use `base_opt.target` in this function.